### PR TITLE
Rework of minimal AI's and citizen's desired activity

### PR DESCRIPTION
**Pull request by someaddons**
# Changes proposed in this pull request:
Rework of our seperated smaller AI's: Eating, Sick, Sleeping, Wander, Flee and Mourning
They are now integrated together in a statemachine, instead of each having their own, 
to allow better communication inbetween and avoids the external synchronization via citizen desired activities
which frequently caused clashes between the minimal AI's and their conditions.

Review please

### CODE REVIEW

File: src/api/java/com/minecolonies/api/entity/ai/ITickingStateAI.java [hunk 1]
```diff
@@ -0,0 +1,11 @@
+package com.minecolonies.api.entity.ai;
+
+/**
+ * AI using our states
+ */
+public interface ITickingStateAI
+{
+    public void tick();
```
Comment (Raycoms) [inline src/api/java/com/minecolonies/api/entity/ai/ITickingStateAI.java:13]: jdoc

File: src/api/java/com/minecolonies/api/entity/ai/statemachine/states/CitizenAIState.java [hunk 2]
```diff
@@ -0,0 +1,19 @@
+package com.minecolonies.api.entity.ai.statemachine.states;
+
+public enum CitizenAIState implements IState
```
Comment (Raycoms) [inline src/api/java/com/minecolonies/api/entity/ai/statemachine/states/CitizenAIState.java:6]: jdoc

File: src/api/java/com/minecolonies/api/entity/ai/statemachine/states/EntityState.java [hunk 3]
```diff
@@ -0,0 +1,13 @@
+package com.minecolonies.api.entity.ai.statemachine.states;
+
+public enum EntityState implements IState
```
Comment (Raycoms) [inline src/api/java/com/minecolonies/api/entity/ai/statemachine/states/EntityState.java:6]: jdoc

File: src/api/java/com/minecolonies/api/entity/citizen/citizenhandlers/ICitizenJobHandler.java [hunk 4]
```diff
@@ -44,4 +45,8 @@ public interface ICitizenJobHandler
      * @return true if so.
      */
     boolean shouldRunAvoidance();
+
+    void setWorkAI(ITickingStateAI workAI);
```
Comment (Raycoms) [inline src/api/java/com/minecolonies/api/entity/citizen/citizenhandlers/ICitizenJobHandler.java:54]: jdoc

File: src/main/java/com/minecolonies/coremod/colony/CitizenData.java [hunk 5]
```diff
@@ -1623,6 +1617,8 @@ public void onGoSleep()
               Component.translatable(CITIZEN_NOT_GUARD_NEAR_HOME),
               ChatPriority.CHITCHAT));
         }
+
+        decreaseSaturation(workBuilding == null || workBuilding.getBuildingLevel() == 0 ? 1 : (SATURATION_DECREASE_FACTOR * Math.pow(2, workBuilding.getBuildingLevel())) * 2);
```
Comment (Raycoms) [inline src/main/java/com/minecolonies/coremod/colony/CitizenData.java:1621]: where did this come from?

File: src/main/java/com/minecolonies/coremod/commands/citizencommands/CommandCitizenInfo.java [hunk 6]
```diff
@@ -110,7 +110,7 @@ else if (entityCitizen.getCitizenColonyHandler().getWorkBuilding() != null && en
                 entityCitizen.getCitizenColonyHandler().getWorkBuilding().getFirstModuleOccurance(WorkerBuildingModule.class).getJobEntry().getTranslationKey()), true);
             context.getSource()
               .sendSuccess(Component.translatable(CommandTranslationConstants.COMMAND_CITIZEN_INFO_ACTIVITY,
-                entityCitizen.getDesiredActivity(),
+                "todo",
```
Comment (Raycoms) [inline src/main/java/com/minecolonies/coremod/commands/citizencommands/CommandCitizenInfo.java]: ?

File: src/main/java/com/minecolonies/coremod/entity/ai/citizen/CitizenAI.java [hunk 7]
```diff
@@ -0,0 +1,264 @@
...
+
+import static com.minecolonies.api.entity.citizen.VisibleCitizenStatus.*;
+import static com.minecolonies.api.research.util.ResearchConstants.WORKING_IN_RAIN;
+import static com.minecolonies.api.util.constant.CitizenConstants.*;
+import static com.minecolonies.api.util.constant.Constants.DEFAULT_SPEED;
+import static com.minecolonies.api.util.constant.TranslationConstants.*;
+import static com.minecolonies.coremod.entity.ai.minimal.EntityAIEatTask.RESTAURANT_LIMIT;
+import static com.minecolonies.coremod.entity.citizen.citizenhandlers.CitizenDiseaseHandler.SEEK_DOCTOR_HEALTH;
+
+public class CitizenAI implements IStateAI
```
Comment (Raycoms) [inline src/main/java/com/minecolonies/coremod/entity/ai/citizen/CitizenAI.java:40]: plenty of jdoc in this class missing

File: src/main/java/com/minecolonies/coremod/entity/ai/citizen/CitizenAI.java [hunk 8]
```diff
@@ -0,0 +1,264 @@
...
+    private       IState        lastState = CitizenAIState.IDLE;
+
+    public CitizenAI(final EntityCitizen citizen)
+    {
+        this.citizen = citizen;
+
+        citizen.getCitizenAI().addTransition(new AIEventTarget<IState>(AIBlockingEventType.EVENT, () -> true, this::decideAiTask, 10));
+        registerWorkAI();
+
+        new EntityAICitizenAvoidEntity(citizen, Monster.class, (float) DISTANCE_OF_ENTITY_AVOID, LATER_RUN_SPEED_AVOID, INITIAL_RUN_SPEED_AVOID);
```
Comment (Raycoms) [inline src/main/java/com/minecolonies/coremod/entity/ai/citizen/CitizenAI.java]: Those are odd, just new? They are no where registered here or so?

File: src/main/java/com/minecolonies/coremod/entity/ai/citizen/CitizenAI.java [hunk 9]
```diff
@@ -0,0 +1,264 @@
...
+        // Eating
+        if (shouldEat())
+        {
+            return CitizenAIState.EATING;
+        }
+
+        // Mourning
+        if (citizen.getCitizenData().getCitizenMournHandler().isMourning() && citizen.getCitizenData().getCitizenMournHandler().shouldMourn())
+        {
+            if (!citizen.getCitizenColonyHandler().getColony().getRaiderManager().isRaided())
```
Comment (Raycoms) [inline src/main/java/com/minecolonies/coremod/entity/ai/citizen/CitizenAI.java]: Ehm, not sure about this? What is this for? Will guards mourn during a raid btw?

File: src/main/java/com/minecolonies/coremod/util/WorkerUtil.java [hunk 10]
```diff
@@ -137,9 +135,7 @@ public static boolean isWorkerAtSiteWithMove(@NotNull final AbstractEntityCitize
         {
             //If not moving the try setting the point where the entity should move to
             if (worker.getNavigation().isDone() && !EntityUtils.tryMoveLivingToXYZ(worker, x, y, z))
-            {
-                worker.getCitizenStatusHandler().setStatus(Status.PATHFINDING_ERROR);
-            }
+            {}
```
Comment (Raycoms) [inline src/main/java/com/minecolonies/coremod/util/WorkerUtil.java]: ??

**Review by Raycoms — COMMENTED** [review]
Some questions and docs

File: src/main/java/com/minecolonies/coremod/entity/ai/citizen/CitizenAI.java (hunk 9 above)
Comment (MotionlessTrain) [reply src/main/java/com/minecolonies/coremod/entity/ai/citizen/CitizenAI.java]: At the moments, guards don't mourn at all

File: src/main/java/com/minecolonies/coremod/entity/ai/citizen/CitizenAI.java (hunk 8 above)
Comment (someaddons) [reply src/main/java/com/minecolonies/coremod/entity/ai/citizen/CitizenAI.java]: it registers itself to the citizens AI statemachine, I showed you it a bit back on discord and asked if you think some kind of registration would be used :D

File: src/main/java/com/minecolonies/coremod/colony/CitizenData.java (hunk 5 above)
Comment (someaddons) [reply src/main/java/com/minecolonies/coremod/colony/CitizenData.java:1621]: from citizens desiredactivity, was triggered when it decided to go to sleep

File: src/main/java/com/minecolonies/coremod/entity/ai/citizen/CitizenAI.java (hunk 9 above)
Comment (someaddons) [reply src/main/java/com/minecolonies/coremod/entity/ai/citizen/CitizenAI.java]: ye this is similar to the previous logic, guards are excluded from all this at the top

File: src/main/java/com/minecolonies/coremod/entity/ai/citizen/CitizenAI.java (hunk 8 above)
Comment (Raycoms) [reply src/main/java/com/minecolonies/coremod/entity/ai/citizen/CitizenAI.java]: I would probably move this out of the constructor directly here. Like statemachine.register(new Thingy) Reading the full context it makes more sense now =D

File: src/main/java/com/minecolonies/coremod/entity/ai/citizen/CitizenAI.java (hunk 9 above)
Comment (Raycoms) [reply src/main/java/com/minecolonies/coremod/entity/ai/citizen/CitizenAI.java]: But can this ever reach this in a raid at all?

File: src/main/java/com/minecolonies/coremod/colony/CitizenData.java (hunk 5 above)
Comment (Raycoms) [reply src/main/java/com/minecolonies/coremod/colony/CitizenData.java:1621]: Ah ok

File: src/main/java/com/minecolonies/coremod/entity/ai/citizen/CitizenAI.java (hunk 9 above)
Comment (someaddons) [reply src/main/java/com/minecolonies/coremod/entity/ai/citizen/CitizenAI.java]: you're right raided doesnt reach here

**Pull request merged**